codec-http/src/main/java/io/netty/handler/codec/spdy/SpdyOrHttpChooser.java

```diff
@@ -106,72 +83,95 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
         }
     }
 
-    private boolean initPipeline(ChannelHandlerContext ctx) {
+    private boolean configurePipeline(ChannelHandlerContext ctx) {
         // Get the SslHandler from the ChannelPipeline so we can obtain the
         // SslEngine from it.
         SslHandler handler = ctx.pipeline().get(SslHandler.class);
         if (handler == null) {
             // SslHandler is needed by SPDY by design.
-            throw new IllegalStateException("SslHandler is needed for SPDY");
+            throw new IllegalStateException("cannot find a SslHandler in the pipeline (required for SPDY)");
         }
 
-        SelectedProtocol protocol = getProtocol(handler.engine());
-        switch (protocol) {
-        case UNKNOWN:
-            // Not done with choosing the protocol, so just return here for now,
+        if (!handler.handshakeFuture().isDone()) {
             return false;
+        }
+
+        SelectedProtocol protocol;
+        try {
+            protocol = selectProtocol(handler);
+        } catch (Exception e) {
+            throw new IllegalStateException("failed to get the selected protocol", e);
+        }
+
+        if (protocol == null) {
+            throw new IllegalStateException("unknown protocol");
+        }
+
+        switch (protocol) {
         case SPDY_3_1:
-            addSpdyHandlers(ctx, SpdyVersion.SPDY_3_1);
+            try {
+                configureSpdy(ctx, SpdyVersion.SPDY_3_1);
+            } catch (Exception e) {
+                throw new IllegalStateException("failed to configure a SPDY pipeline", e);
+            }
             break;
         case HTTP_1_0:
         case HTTP_1_1:
-            addHttpHandlers(ctx);
+            try {
+                configureHttp1(ctx);
+            } catch (Exception e) {
+                throw new IllegalStateException("failed to configure a HTTP/1 pipeline", e);
+            }
             break;
-        default:
```

Comment:
I think keep a default is a good thing so we catch errors if we ever change stuff again